### ISSUE TITLE
Why is `preredering` from `$app/env` always `false`, even during `pnpm build`?

### ISSUE THREAD

**Issue by ar4hc** [sveltejs/kit#4265]
### Describe the bug

i use hooks.js to init a redis connection.
And in the build process hooks gets called and tries to connect to redis.
Which is in a docker container and far away and not availabe during build, 
so it breaks and crashes the build.

So i tried `prerendering`, and return something "empty" but this fails too as it is always reported as `false`...

My code looks like this:

```js
import { prerendering } from '$app/env';

console.info('Prerendering:', prerendering);

const client = createClient(
  // [...]
);

export const handle = async ({ event, resolve }) => {
		if (!client.isOpen) await client.connect();
  // [...]
}
```

and in handle() i'd try to wrap some of that code in `if (prerendering)...`, and avoid the db connection.

i'm not sure i understand how `prerendering` and hooks should work here, but i think it should work that way...

### Reproduction

run pnpm build in shell, `docker build`, or `docker compose build`

### Logs

```shell
$ pnpm build

> classifier-gui@0.0.1 build /Users/ar/src/lincare/classifier/classifier-gui
> svelte-kit build

vite v2.8.6 building for production...
✓ 54 modules transformed.
.svelte-kit/output/client/_app/manifest.json                               2.11 KiB
.svelte-kit/output/client/_app/pages/__layout.svelte-687f8954.js           2.00 KiB / gzip: 1.06 KiB
.svelte-kit/output/client/_app/pages/__error.svelte-6723fd33.js            3.69 KiB / gzip: 1.20 KiB
[...]
vite v2.8.6 building SSR bundle for production...
✓ 62 modules transformed.
.svelte-kit/output/server/manifest.json                                    3.27 KiB
.svelte-kit/output/server/index.js                                         67.47 KiB
.svelte-kit/output/server/entries/endpoints/index.js                       2.17 KiB
.svelte-kit/output/server/entries/endpoints/faxlist.js                     1.17 KiB
.svelte-kit/output/server/entries/endpoints/api/index.js                   0.07 KiB
[...]

Prerendering: false <--- ouput from the console.log above...
```

### System Info

```shell
System:
    OS: macOS 11.6.4
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
    Memory: 1.43 GB / 32.00 GB
    Shell: 5.8.1 - /usr/local/bin/zsh
  Binaries:
    Node: 17.6.0 - /usr/local/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 8.5.1 - /usr/local/bin/npm
    Watchman: 2022.02.28.00 - /usr/local/bin/watchman
  Browsers:
    Firefox: 97.0.1

But also in docker build on linux node base. image

and i'm using pnpm.
```

### Severity

serious, but I can work around it

### Additional Information

More a question than a bug.

**Comment by Christian-E** [comment]
Hey! I stumbled upon the same issue yesterday when I tried initializing a mongodb connection using a top-level await.
I have the same understanding about how `prerendering` should work.
For now I was able to work around that by globally disabling prerendering in my svelte.config.js (as it isn't needed for now) - but obviously I would prefere enabling this again at some point.

**Comment by Theo-Steiner** [comment]
Prerendering is always false here, since you access it here before SvelteKit has the opportunity to set it to true.
First here's a quick workaround that should work for your use case:
```javascript
// inside hooks.js
import { prerendering } from '$app/env';

// At this point, prerendering will always be false due to SvelteKit's implementation
let client;

function setUpClient() {
	if (client || prerendering) return;
	client = createClient();
}

export const handle = async ({ event, resolve }) => {
        // But once you are inside the handle hook (Or any other hook) you'll have access to the correct value of prerendering
	setUpClient();
	if (client) doHereWhatYouCanOnlyDoHere();
	const response = await resolve(event);
	return response;
};
```

Now let's get into why this doesn't work:
Everything on the toplevel of hooks.js is the global code of that module. Meaning that it will run as soon as the module "hooks.js" is imported. This happens (as you might have guessed) [during the prerender step of the build process.](https://github.com/sveltejs/kit/blob/7f958dd49cfbbc4712001b1b21ce1474b44da69e/packages/kit/src/core/build/prerender/prerender.js#L67)
```javascript
// Line 67 of prerender.js
	const { Server, override } = await import(pathToFileURL(`${server_root}/server/index.js`).href);
	const { manifest } = await import(pathToFileURL(`${server_root}/server/manifest.js`).href);
// Line 70
	override({
		paths: config.kit.paths,
		prerendering: true,
		read: (file) => readFileSync(join(config.kit.files.assets, file))
	});
```
So the global code of your hooks.js file is invoked on line 67 and at that time "prerendering" is still at the value of "false", [with which it was initialized.](https://github.com/sveltejs/kit/blob/1d76d66038647c8a67dbb011cf9c0e231fd086a6/packages/kit/src/runtime/env.js)
On line 70 of prerender.js prerendering is then set to the correct value, which is why from that point forward (for example whenever the hook is actually called) we have access to the actual "prerendering" state.

Now, why does the override function to set the prerendering state and the hooks import end up mambo-jamboed into the same server/index.js file? 
This I don't really know...
The build process [calls the build_server function](https://github.com/sveltejs/kit/blob/7f958dd49cfbbc4712001b1b21ce1474b44da69e/packages/kit/src/core/build/index.js#L46) in which this single server/index.js is generated from a template and written to the file system at a path specified by the user. As only prerender seems to ever access the override function exported from that module, I wonder if this behavior could be fixed simply by not writing override to server.js and invoking it from either the main build script or the prerender script before the problematic import of server/index.js...

**Comment by Rich-Harris** [comment]
Closed via #4322

**Comment by ar4hc** [comment]
Thanks for the explainer and the fix.
